sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/BulkWriter.scala

```diff
@@ -130,6 +143,202 @@ class BulkWriter(container: CosmosAsyncContainer,
+          .getMaxMicroBatchInterval(cosmosBulkExecutionOptions)
+
+      readManyInputEmitterOpt
+          .get
+          .asFlux()
+          .publishOn(readManyBoundedElastic)
+          .bufferTimeout(bulkBatchSize, batchInterval)
+          .flatMap(readManyOperations => {
+              val cosmosIdentitySet = readManyOperations.map(option => option.cosmosItemIdentity).toSet
+
+              // for each batch, use readMany to read items from cosmosdb
+              container
+                  .readMany(cosmosIdentitySet.toList.asJava, classOf[ObjectNode])
+                  .switchIfEmpty(
+                      Mono.just(
+                          ImplementationBridgeHelpers
+                              .FeedResponseHelper
+                              .getFeedResponseAccessor
+                              .createFeedResponse(new util.ArrayList[ObjectNode](), null, null)))
+                  .doOnNext(feedResponse => {
+                      val resultMap = new TrieMap[CosmosItemIdentity, ObjectNode]()
+                      for (itemNode: ObjectNode <- feedResponse.getResults.asScala) {
+                          resultMap += (
+                              new CosmosItemIdentity(
+                                  PartitionKeyHelper.getPartitionKeyPath(itemNode, partitionKeyDefinition),
+                                  itemNode.get(CosmosConstants.Properties.Id).asText()) -> itemNode)
+                      }
+
+                      // It is possible that multiple cosmosPatchBulkUpdateOperations were targeting for the same item
+                      // Currently, we are still creating one bulk item operation for each cosmosPatchBulkUpdateOperations
+                      // for easier exception and semaphore handling
+                      // However a consequences of it could be, the generated bulk item operation will fail due to conflicts or pre-condition failure
+                      // If this turns out to be a problem, we can do more optimization here: merge multiple cosmosPatchBulkUpdateOperations into one bulkItemOperation
+                      // But even the above approach can only work within the same batch but not for the whole spark partition processing.
+                      for (readManyOperation <- readManyOperations) {
+                          val cosmosPatchBulkUpdateOperations =
+                              cosmosPatchHelperOpt
+                                  .get
+                                  .createCosmosPatchBulkUpdateOperations(readManyOperation.objectNode)
+
+                          val rootNode =
+                              cosmosPatchHelperOpt
+                                  .get
+                                  .patchBulkUpdateItem(resultMap.get(readManyOperation.cosmosItemIdentity), cosmosPatchBulkUpdateOperations)
+
+                          // create bulk item operation
+                          val etagOpt = Option.apply(rootNode.get(CosmosConstants.Properties.ETag))
```

Comment:
Question: Is this done to match Spark 2 - would we allow to configure whether to use etag precondition always (I see why it makes sense - so no criticism - just trying to fully understand the intend here)

Reply:
Never mind - etag definitely is needed - you just need to make sure the 409/412 handling includes the read-many on retry